contribute/design/proposal-cstor-orchestration.md

```diff
@@ -0,0 +1,376 @@
+
+
+### Design Considerations
+The previous two sections have laid out the workflow for a successful pool and volume creation. As part of the workflow, several cases need to be considered like:
+- Node hosting the CStorPool is down or not reachable. 
+- Node hosting the cStorContainer is running out of resources and cStorContainer is evicted
+- Node hosting the CStorPool is restarted
+- OpenEBS Volume with replica count = 1 and CStorPool is restarted. 
+- OpenEBS Volume with replica count = 3 and case where 1 of the 3 replica nodes, 2 of 3 replica nodes and 3 or 3 replica nodes are down
+- All nodes are down and are restarted one by one
+- OpenEBS Volume (PVC) is accidentally deleted and user wants to get the data stored in the volume back. 
+- OpenEBS Volume data needs to be backed up or restored from a backup. 
+- CStorPool has a capacity of 100G and volumes are created adding up to more than 100G
+- One of the disks of the CStorPool is showing high latency
+- 
+
+
+
+## Implementation Plan
+
+### Phase 1
+- Install/Setup the CRDs used in this design
+- Container images for - cstor-pool, cstor-pool-mgmt
+- cstor-pool-mgmt sidecar interfaces between observing the CStorPool CR objects and issues - pool create and delete
+
+### Phase 2
+- Container images for - cstor-ctrl, cstor-ctrl-mgmt
+- cstor-pool-mgmt sidecar interfaces between observing the CStorVolumeReplica CR objects and issues - volume create and delete
+- cstor-ctrl-mgmt sidecar interfaces between observing the CStorVolume CR objects and generates the configuration required for cstor-ctrl
+- Enhance the maya-exporter to interface with cstor-pool to gather pool level metrics 
+- Enhance the maya-exporter to interface with cstor-ctrl to gather volume level metrics 
+
+### Phase 3
+- Enhance the maya-apiserver to observer StoragePoolClaim and create CStorPool, StoragePool and associated Kubernetes Deployments and Services
+- Enhance the maya-apiserver to create cstor based volumes - which will involve creating CStorVolumeReplica, CStorVolume and associated Kubernetes Deployments and Services.
+- Grafana Dashboard for showing the cStor Pool status and statistics
+- Grafana Dashboard for showing the cStor Volume status and statistics
+
+
+### Future
+- Upgrade of cstor related containers
+- Editing either the Pool or Volume related parameters
+- Replacing failed disks using a spare disk from the pool
+- Expanding the pool to add more capacity
+- Marking a Pool as unavailable or failed due to slow disk or to bring it down for maintenance of the underlying disks
+- Reassign the pool from one node to another by shifting the attached disks to a new node
+- User should be able to specify required values for the features available on the CStorPool and CStorVolumeReplica like compression, block size, deduplication, etc. This has to be aligned with the VolumePolicies and VolumeUpdatePolicies being implemented in OpenEBS 0.6 and 0.7 respectively.
+
+
```

Comment:
Any plans of having volume QOS policies? Will these be part of VolumePolicies mentioned above?

Reply:
yes @avishnu the QoS policies will be implemented by VolumePolicies - that will inturn translated into resource settings on the deployment yamls. We may also in the future consider setting these values into the Pool, VolumeController and VolumeReplica CRs for finer control. Added this as a future item.